tests/unittests/CoreGraphics.drawing/ImageComparison.h

```diff
@@ -39,7 +41,107 @@ class ImageComparator {
     virtual ImageDelta CompareImages(CGImageRef left, CGImageRef right) = 0;
 };
 
+struct bgraPixel {
+    uint8_t b, g, r, a;
+};
+
+struct rgbaPixel {
+    uint8_t r, g, b, a;
+};
+
+template <typename T, typename U>
+bool operator==(const T& t, const U& u) {
+    return t.r == u.r && t.g == u.g && t.b == u.b && t.a == u.a;
+}
+
+template <ComparisonMode Mode, size_t FailureThreshold = 1>
 class PixelByPixelImageComparator : public ImageComparator {
 public:
-    ImageDelta CompareImages(CGImageRef left, CGImageRef right) override;
+    ImageDelta CompareImages(CGImageRef left, CGImageRef right) override {
+        if (!left || !right) {
+            return { ImageComparisonResult::Incomparable };
+        }
+
+        CGSize leftSize{
+            (CGFloat)CGImageGetWidth(left), (CGFloat)CGImageGetHeight(left),
+        };
+
+        CGSize rightSize{
+            (CGFloat)CGImageGetWidth(right), (CGFloat)CGImageGetHeight(right),
+        };
+
+        size_t leftPixelCount = leftSize.width * leftSize.height;
+        size_t rightPixelCount = rightSize.width * rightSize.height;
+
+        if (leftPixelCount != rightPixelCount) {
+            return { ImageComparisonResult::Incomparable };
+        }
+
+        woc::unique_cf<CFDataRef> leftData{ _CFDataCreateFromCGImage(left) };
+        woc::unique_cf<CFDataRef> rightData{ _CFDataCreateFromCGImage(right) };
+
+        CFIndex leftLength = CFDataGetLength(leftData.get());
+        if (leftLength != CFDataGetLength(rightData.get())) {
+            return { ImageComparisonResult::Incomparable };
+        }
+
+        woc::unique_iw<uint8_t> deltaBuffer{ static_cast<uint8_t*>(IwCalloc(leftLength, 1)) };
+
+        const rgbaPixel* leftPixels{ reinterpret_cast<const rgbaPixel*>(CFDataGetBytePtr(leftData.get())) };
+        const rgbaPixel* rightPixels{ reinterpret_cast<const rgbaPixel*>(CFDataGetBytePtr(rightData.get())) };
+        rgbaPixel* deltaPixels{ reinterpret_cast<rgbaPixel*>(deltaBuffer.get()) };
+
+        // ASSUMPTION: The context draw did not cover the top left pixel;
+        // We can use it as the background to detect accidental background deletion and miscomposition.
+        auto background = leftPixels[0];
+
+        size_t npxchg = 0;
+        for (off_t i = 0; i < leftLength / sizeof(rgbaPixel); ++i) {
+            auto& bp = leftPixels[i];
+            auto& cp = rightPixels[i];
+            auto& gp = deltaPixels[i];
+            if (!(bp == cp)) {
+                ++npxchg;
+                if (cp == background) {
+                    // Pixel is in EXPECTED but not ACTUAL
+                    gp.r = gp.a = 255;
+                } else if (bp == background) {
+                    // Pixel is in ACTUAL but not EXPECTED
+                    gp.g = gp.a = 255;
+                } else {
+                    // Pixel is in BOTH but DIFFERENT
+                    if (Mode == ComparisonMode::Mask) {
```

Comment:
>Mode [](start = 24, length = 4)

Not the right way to do this.  You want comparison traits, which does this.  This would be ComparisonTraits::ComputeDelta() where this entire work will be done.